src/node_v8.cc

```diff
@@ -210,6 +211,173 @@ void SetFlagsFromString(const FunctionCallbackInfo<Value>& args) {
+    writer->json_keyvalue("spaceAvailableSize",
+                          heap_space_statistics.space_available_size());
+    writer->json_keyvalue("physicalSpaceSize",
+                          heap_space_statistics.physical_space_size());
+    writer->json_end();
+  }
+  writer->json_arrayend();
+}
+
+static void BeforeGCCallback(Isolate* isolate,
+                             v8::GCType gc_type,
+                             v8::GCCallbackFlags flags,
+                             void* data) {
+  GCProfiler* profiler = static_cast<GCProfiler*>(data);
+  if (profiler->current_gc_type() != 0) {
+    return;
+  }
+  profiler->set_current_gc_type(gc_type);
+  profiler->set_start_time(uv_hrtime());
+  JSONWriter* writer = profiler->writer();
+  writer->json_start();
+  writer->json_keyvalue("gcType", GetGCTypeName(gc_type));
+  writer->json_objectstart("beforeGC");
+  SetHeapStatistics(writer, isolate);
+  writer->json_objectend();
+}
+
+static void AfterGCCallback(Isolate* isolate,
+                            v8::GCType gc_type,
+                            v8::GCCallbackFlags flags,
+                            void* data) {
+  GCProfiler* profiler = static_cast<GCProfiler*>(data);
+  if (profiler->current_gc_type() != gc_type) {
+    return;
+  }
+  JSONWriter* writer = profiler->writer();
+  profiler->set_current_gc_type(0);
+  u_int64_t start_time = profiler->start_time();
+  profiler->set_start_time(0);
+  writer->json_keyvalue("cost", (uv_hrtime() - start_time) / 1e3);
+  writer->json_objectstart("afterGC");
+  SetHeapStatistics(writer, isolate);
+  writer->json_objectend();
+  writer->json_end();
+}
+
+GCProfiler::GCProfiler(Environment* env, Local<Object> object)
+    : BaseObject(env, object), writer_(outfile_, false) {
+  MakeWeak();
+}
+
+// This function will be called when
+// 1. StartGCProfile and StopGCProfile are called and
+//    JS land do not keep the object any more.
+// 2. StartGCProfile is called then the env exits before
+//    StopGCProfile is called.
+GCProfiler::~GCProfiler() {
+  if (state_ != GCProfilerState::kInitialized) {
+    env()->isolate()->RemoveGCPrologueCallback(BeforeGCCallback, this);
+    env()->isolate()->RemoveGCEpilogueCallback(AfterGCCallback, this);
+  }
+}
+
+void GCProfiler::New(const FunctionCallbackInfo<Value>& args) {
+  CHECK(args.IsConstructCall());
+  Environment* env = Environment::GetCurrent(args);
+  new GCProfiler(env, args.This());
+}
+
+void GCProfiler::StartGCProfile(const FunctionCallbackInfo<Value>& args) {
+  GCProfiler* profiler;
+  ASSIGN_OR_RETURN_UNWRAP(&profiler, args.Holder());
+  if (profiler->state() != GCProfilerState::kInitialized) {
+    return;
+  }
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = args.GetIsolate();
+  node::Utf8Value filename(env->isolate(), args[0]);
+  profiler->outfile()->open(*filename, std::ios::out | std::ios::binary);
+  if (!profiler->outfile()->is_open()) {
+    env->ThrowError("failed to open file");
```

Comment:
This should probably result in an actual error message, but generally speaking, what I’m missing here is _why_ we would write to the filesystem instead of e.g. returning the result in a callback/promise? The caller could still write it to disk if they want to.

Reply:
I agree with @addaleax . I don't think we should write it to disk.

Reply:
Thanks, i have updated the code, the API will return an object to the user.